### PR TITLE
30% optimization of `DateTime.GetDate()`/`.Year/.Month/.Day/.DayOfYear` by 'Euclidean affine functions'

### PR DESCRIPTION
**Pull request by SergeiPavlov**
This optimization based on article  
[Cassio Neri, Lorenz Schneider - Euclidean Affine Functions and Applications to Calendar Algorithms - 2021](https://arxiv.org/pdf/2102.06959.pdf)

It avoids loops and minimizes branching in calendar calculations.

Benchmarks:
```
|                                                     Method |      Mean | Error |       Min |       Max | Allocated |
|----------------------------------------------------------- |----------:|------:|----------:|----------:|----------:|
|           'DateTime.GetDate(out year, out month, out day)' |  7.697 ns |    NA |  7.697 ns |  7.697 ns |         - |
| 'Optimized DateTime.GetDate(out year, out month, out day)' |  5.114 ns |    NA |  5.114 ns |  5.114 ns |         - |


|                     Method |     Mean | Error |      Min |      Max | Allocated |
|--------------------------- |---------:|------:|---------:|---------:|----------:|
|              DateTime.Year | 4.289 ns |    NA | 4.289 ns | 4.289 ns |         - |
|             DateTime.Month | 6.534 ns |    NA | 6.534 ns | 6.534 ns |         - |
|               DateTime.Day | 6.946 ns |    NA | 6.946 ns | 6.946 ns |         - |
|  'Optimized DateTime.Year' | 2.945 ns |    NA | 2.945 ns | 2.945 ns |         - |
| 'Optimized DateTime.Month' | 3.161 ns |    NA | 3.161 ns | 3.161 ns |         - |
|   'Optimized DateTime.Day' | 3.638 ns |    NA | 3.638 ns | 3.638 ns |         - |

```

It is difficult to explain values of magic constants of the algorithm in comments.
They are inferred algebraically.

* Get rid of `GetDatePart()` function and inline its functionality directly into `Year, Month, Day, DayOfYear` properties

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs [hunk 1]
```diff
@@ -1351,92 +1352,58 @@ public DateTime Date
...
-            // Last year has an extra day, so decrement result if 4
-            if (y1 == 4) y1 = 3;
-            // If year was requested, compute and return it
-            if (part == DatePartYear)
+            // y400 = number of whole 400-year periods since 3/1/0000
+            // r1 = day number within 400-year period
+            (uint y400, uint r1) = Math.DivRem(((uint)(UTicks / TicksPer6Hours) | 3U) + 1224, DaysPer400Years);
+            ulong u2 = (ulong)Math.BigMul(2939745, (int)r1 | 3);
+            ushort daySinceMarch1 = (ushort)((uint)u2 / 11758980);
+            var n3 = 2141 * daySinceMarch1 + 197913;
```
Comment (Clockwork-Muse) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1365]: ... Some additional comments, especially around the specific magic numbers, would be greatly appreciated.

Comment (SergeiPavlov) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1365]: These magic number have not explainable meaning. They just give right result.
To understand them it is neccessary to follow chain of propisitions in the article.

Comment (Clockwork-Muse) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1365]: Then you should at least be noting the source inside the code.  If there is any sort of naming for those constants in the source, you should probably be pulling them out into named constants as well.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs [hunk 2]
```diff
@@ -1351,92 +1352,58 @@ public DateTime Date
...
+            switch (part)
+            {
+                case DatePartDay:
+                    return (ushort)n3 / 2141 + 1;
+                case DatePartMonth:
+                    return (ushort)(n3 >> 16) - (daySinceMarch1 >= 306 ? 12 : 0);
+            }
+
+            var year = (int)(100 * y400 + (uint)(u2 >> 32)) + (daySinceMarch1 >= 306 ? 1 : 0);
+            switch (part)
```
Comment (MichalPetryka) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1375]: This should probably just be an if?

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs (hunk 1 above)
Comment (SergeiPavlov) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1365]: I did in the function comment https://github.com/dotnet/runtime/pull/72712/files#diff-5afffb30d8c00fa852b7df6e5bce6255be93acc1281d20b308ab57e8b70f567bR1356

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs (hunk 2 above)
Comment (SergeiPavlov) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1375]: > This should probably just be an if?

`if` would be terse, but as `part` is switching key between 4 different cases, it is reasonable to use `switch` statement, even while its parts are scattered over function code

Comment (teo-tsirpanis) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1375]: As it is now the code will take advantage of dotnet/roslyn#61483 once implemented.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs [hunk 3]
```diff
@@ -1351,92 +1352,58 @@ public DateTime Date
...
-            while (n >= days[m]) m++;
-            // compute month and day
-            month = (int)m;
-            day = (int)(n - days[m - 1] + 1);
+            // y400 = number of whole 400-year periods since 3/1/0000
+            // r1 = day number within 400-year period
+            (uint y400, uint r1) = Math.DivRem(((uint)(UTicks / TicksPer6Hours) | 3U) + 1224, DaysPer400Years);
+            ulong u2 = (ulong)Math.BigMul(2939745, (int)r1 | 3);
+            ushort daySinceMarch1 = (ushort)((uint)u2 / 11758980);
+            var n3 = 2141 * daySinceMarch1 + 197913;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/DateTime.cs:1396]: Nit: Our [coding conventions](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md) disallow vars, except in a few limited cases. These should be explicit types.

**Comment by danmoseley** [comment]
Do we have test coverage here sufficient to give high confidence? If you can think of any other useful test cases this PR is a good time to add them.

**Comment by danmoseley** [comment]
Maybe @lemire is interested in this trick.

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs (hunk 2 above)
Comment (SergeiPavlov) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: changed to `if`

File: src/libraries/System.Private.CoreLib/src/System/DateTime.cs (hunk 3 above)
Comment (SergeiPavlov) [reply src/libraries/System.Private.CoreLib/src/System/DateTime.cs]: fixed it

**Comment by SergeiPavlov** [comment]
> Do we have test coverage here sufficient to give high confidence? If you can think of any other useful test cases this PR is a good time to add them.

I tested it for every possible date (range 1/1/0001 .. 12/31/9999). Results are identical to current implementation.
Of course it does not replace testing in dotnet workflow.

**Comment by lemire** [comment]
@danmoseley I knew of the article, but I am happy to see this work come to dotnet.

**Comment by danmoseley** [comment]
I've compared the paper to the code changes here and I can reproduce the calculations, so the code makes sense to me.  (Although, I see `r2 = u2 % 2^32 / 2939745 / 4`, is turning into `(ushort)((uint)u2 / 11758980)` .. clearly produces the right result, but I would think precedence would run the divisions one after the other, rather than multiplying .. what am I missing about their notation..)

I added some more random tests locally and they all pass, which should not be a surprise since you say you've tried all possible valid dates yourself.

> Adding [MethodImpl(MethodImplOptions.AggressiveInlining)] to GetDatePart() makes properties Year, Month, Day, DayOfYear significantly more efficient. The compiler discards unrelated branches of GetDatePart().

Would it make for clearer code, albeit with a bit more duplication, to manually inline the appropriate parts inside each of these? Then it will be self evident that the unrelated branches are discarded.

Re my question about perf tests, I guess this covers it indirectly. Did you run this test before/after? But, it would probably be good to add benchmarks in this file for Year, Month, Day, DayOfYear individually, especially if you end up copy/pasting instead of inlining as suggested above.
https://github.com/dotnet/performance/blob/e60e729d0008d8522b518b11c06b9803f6d674ac/src/benchmarks/micro/libraries/System.Runtime/Perf.DateTime.cs#L34

**Comment by danmoseley** [comment]
@tannergooding in case he is interested in the math..

**Comment by danmoseley** [comment]
I notice we are missing tests for DayOfYear that use randomized inputs. Year, Month, Day are indirectly covered by the ToString() tests with randomized inputs
https://github.com/dotnet/runtime/blob/cc7ccfd0763b1f3dca173268ac7295ab4afeefb8/src/libraries/System.Runtime/tests/System/DateTimeTests.cs#L2149
Those ToString() tests indirectly test Year, Month, Day getters.

**Comment by danmoseley** [comment]
@richlander when we're using an algorithm from a paper like this, ie., there's no particular software license, I assume credit in a comment is fine, and no entry in THIRD-PARTY-NOTICES.TXT is expected?

**Comment by SergeiPavlov** [comment]
> Would it make for clearer code, albeit with a bit more duplication, to manually inline the appropriate parts inside each of these? Then it will be self evident that the unrelated branches are discarded.

Inlined them

**Comment by SergeiPavlov** [comment]
> Re my question about perf tests, I guess this covers it indirectly. Did you run this test before/after? But, it would probably be good to add benchmarks in this file for Year, Month, Day, DayOfYear individually, especially if you end up copy/pasting instead of inlining as suggested above. (If you add benchmarks, that would of course be a PR into dotnet/performance that we'd merge after code flows, but you could run them locally and report here.)

Locally I have a source file with both: original & new implementations. The first benchmarking result were before applying `[MethodImpl(MethodImplOptions.AggressiveInlining)]`.

Added individual benchmarks for the properties. Results ar in the PR description. The most significant gain (50%) is for `Month` which was most heavy because of loop.

**Review by danmoseley — APPROVED** [review]
LGTM - I didn't recheck the math in detail, if the tests pass.

As mentioned, DayOfYear coverage is relatively poor, while the others have good indirect coverage, it would be nice to add some tests with randomized data.

Thanks for the nice improvement!

**Review by danmoseley — APPROVED** [review]
on latest changes

**Comment by danmoseley** [comment]
@richlander is going to check on licensing here. Marking **do not merge** meantime. The change is otherwise fine.

**Comment by richlander** [comment]
Right. I sent mail to the authors of the paper + @SergeiPavlov.

**Comment by cassioneri** [comment]
Lorenz and I are glad to see that our algorithm might go into .NET. We wish to thank, especially, Sergei for implementing and creating the PR.

On the licensing question sent by email, Lorenz and I have discussed the matter and we agreed with your second bullet point which I quote:

> Agree that the code was primarily informed by the paper and that it is licensed permissively such that we can land the PR.

It is clear that Sergei did implement the algorithm as explained in the paper. In terms of implementation, we consider his C# code substantially different from the GPL-licensed C++ code found in my repository.

For the reasons above, as far as Lorenz and I are concerned, we do not believe that GPL applies to Sergei's PR. As I said above, we would like to see the PR merged into the .NET's main branch for the benefit of .NET users. We would appreciate (but not require) if the release notes could mention our work.

**Comment by danmoseley** [comment]
Appreciation to @cassioneri for the licensing. I will follow up with an addition to the third party notices, no need to reset CI.

AOT and tvOS timed out, but previously passed, and vanishingly likely to be relevant here.

**Comment by richlander** [comment]
We will add in release notes and a blog post (including the release blog post). Thanks much!

@JonDouglas

**Comment by IDisposable** [comment]
> I've compared the paper to the code changes here and I can reproduce the calculations, so the code makes sense to me. (Although, I see `r2 = u2 % 2^32 / 2939745 / 4`, is turning into `(ushort)((uint)u2 / 11758980)` .. clearly produces the right result, but I would think precedence would run the divisions one after the other, rather than multiplying .. what am I missing about their notation..)

Ignoring the `% 2^32` part... the basic math rule at hand here is that `X / Y / Z === X / (Y * Z)`. The easiest way to grok this is an example:  64 / 4 / 4 -> left-to-right apply first divisions `(64 / 4) / 4 == 16 / 4` -> apply remaining division `16 / 4 == 4`... now let's pre-apply the second division (which can be thought of as _scaling up_ the divisor of the first one... so `64 / ( 4 * 4) == 64 / 16` and then apply the first division so `64 / 16 == 4`... same answer... and if we KNOW the right-side divisors are constant there's no reason not to pre-multiply.

**Comment by lemire** [comment]
+1

**Comment by SergeiPavlov** [comment]
> `((X / Y) / Z) * Z = a * Z` --> move over the `Z` by multiplying both sides by `Z`
> `(X / Y) = a * Z` --> which isolates `X / Y` on left

Hm, this implication does not work for integer division in general case.
Counter-example:
```
X = 21
Y = 3
Z = 2
a = X / Y / Z = 3
```
`((21 / 3) / 2) * 2` = `3 * 2`
but
`(21 / 3)` ≠ `3 * 2`

**Comment by IDisposable** [comment]
Certainly there are truncation errors for integers where significance is lost, that's why I was talking about it in algebraic form :)

**Comment by SergeiPavlov** [comment]
The proof for integer division:

let   X = YA+B;     X/Y = A = ZQ+R
where Y = (B+M), 1 <= M
      Z = (R+N), 1 <= N

then:
YZ = (B+M)(R+N) = BR+BN +MR+MN

X = YA + B = Y(ZQ+R) + B = YZQ + YR + B = (YZ)Q + (B+M)R + B = (YZ)Q + BR + MR + B

X/(YZ) = ((YZ)Q + BR + MR + B) / YZ 
 = Q + (BR+MR+B)/YZ
 = Q + (BR+MR+B)/(BR+BN+RM+MN) = Q       because  BR+MR+B < BR+BN+RM+MN always

**Comment by SergeiPavlov** [comment]
> `((Y * (X / Y)) / Z = a * Y` --> move over the Y by multiplying both sides by Y
> `(X / Z) = a * Y` --> which isolates X / Z on left

again `(Y * (X / Y)`  is not always the same as `X`
Rational numbers algebra rules are not applicable for integer arithmetics.

**Comment by cassioneri** [comment]
FWIW:
1. I agree with Sergei's proof provided that `Y > 0` and `Z > 0`. (Otherwise, here is a counter-example: `X = 10`, `Y = -3`, `A = -3`, `B = 1`, `M = -4`, that is, `1 <= M` is false.)
2. For `X < 0` or `Z < 0`, the result `X/(YZ) = X/Y/Z` is still valid in C and C++ (where `/` means truncation towards 0) and the proof only involves the result already proven for `X > 0` and `Z > 0` and playing with sign rules of multiplication/division.
3. I don't know the exactly specification of `/` in C# but I bet that for positive operands they are the same as in C and C++ and that's all that matters in our case.
4. [Here](https://godbolt.org/z/T7M8Pqq1e) we can see that _some_ C compilers can figure out the optimization and replace `u2 / 2939745 / 4` with `u2 / 11758980`. That's why we did not do it in the paper -- we trusted our compiler would do it.
5. Unfortunately, the C# compiler can't do the same. (See [here](https://godbolt.org/z/KqzTfrGEd) that in `f` there are 2 consecutive `shr`s -- right shifts -- that could be collapsed into a single one but it's not.) Hence the optimization must be done manually. (As Sergei did.)

**Comment by EgorBo** [comment]
Perf improvements https://github.com/dotnet/perf-autofiling-issues/issues/6998

image

Nice!

**Comment by DrewScoggins** [comment]
Also improvements on 
win-x64: https://github.com/dotnet/perf-autofiling-issues/issues/7174

**Comment by kunalspathak** [comment]
https://github.com/dotnet/perf-autofiling-issues/issues/7279

**Comment by danmoseley** [comment]
> Unfortunately, the C# compiler can't do the same. (See [here](https://godbolt.org/z/KqzTfrGEd) that in f there are 2 consecutive shrs -- right shifts -- that could be collapsed into a single one but it's not.) Hence the optimization must be done manually. (As Sergei did.)

Opened an issue for this. Thanks everyone for explaining the math :)

**Pull request merged**